mapping/mapping_vectors.go

```diff
@@ -36,45 +36,97 @@ func NewVectorFieldMapping() *FieldMapping {
-	pathString string, path []string, indexes []uint64, context *walkContext) {
-	propertyVal := reflect.ValueOf(propertyMightBeVector)
-	if !propertyVal.IsValid() {
-		return
+// validate and process a flat vector
+func processFlatVector(vecI interface{}, dims int) ([]float32, bool) {
+	vecV := reflect.ValueOf(vecI)
+	if !vecV.IsValid() || vecV.Kind() != reflect.Slice || vecV.Len() != dims {
+		return nil, false
+	}
+
+	rv := make([]float32, dims)
+	for i := 0; i < vecV.Len(); i++ {
+		item := vecV.Index(i)
+		if !item.CanInterface() {
+			return nil, false
+		}
+		itemI := item.Interface()
+		itemFloat, ok := util.ExtractNumericValFloat32(itemI)
+		if !ok {
+			return nil, false
+		}
+		rv[i] = itemFloat
+	}
+
+	return rv, true
+}
+
+// validate and process a vector
+// max supported depth of nesting is 2 ([][]float32)
+func processVector(vecI interface{}, dims int) ([]float32, bool) {
+	vecV := reflect.ValueOf(vecI)
+	if !vecV.IsValid() || vecV.Kind() != reflect.Slice || vecV.Len() == 0 {
+		return nil, false
+	}
+
+	// Let's examine the first element (head) of the vector.
+	// If head is a slice, then vector is nested, otherwise flat.
+	headI := vecV.Index(0).Interface()
+	headV := reflect.ValueOf(headI)
+	if !headV.IsValid() {
+		return nil, false
+	}
+	if headV.Kind() != reflect.Slice { // vector is flat
+		return processFlatVector(vecI, dims)
 	}
 
-	// Validating the length of the vector is required here, in order to
-	// help zapx in deciding the shape of the batch of vectors to be indexed.
-	if propertyVal.Kind() == reflect.Slice && propertyVal.Len() == fm.Dims {
-		vector := make([]float32, propertyVal.Len())
-		isVectorValid := true
-		for i := 0; i < propertyVal.Len(); i++ {
-			item := propertyVal.Index(i)
-			if item.CanInterface() {
-				itemVal := item.Interface()
-				itemFloat, ok := util.ExtractNumericValFloat32(itemVal)
-				if !ok {
-					isVectorValid = false
-					break
-				}
-				vector[i] = itemFloat
-			}
+	// # process nested vector
+
+	// pre-allocate memory for the flattened vector
+	// so that we can use copy() later
+	rv := make([]float32, dims*vecV.Len())
+
+	for i := 0; i < vecV.Len(); i++ {
+		subVec := vecV.Index(i)
+		if !subVec.CanInterface() {
+			return nil, false
 		}
-		// Even if one of the vector elements is not a float32, we do not index
-		// this field and return silently
-		if !isVectorValid {
-			return
+		subVecI := subVec.Interface()
+		subVecV := reflect.ValueOf(subVecI)
+		if !subVecV.IsValid() {
+			return nil, false
+		}
+
+		if subVecV.Kind() != reflect.Slice {
+			return nil, false
 		}
 
-		fieldName := getFieldName(pathString, path, fm)
-		options := fm.Options()
-		field := document.NewVectorFieldWithIndexingOptions(fieldName,
-			indexes, vector, fm.Dims, fm.Similarity, options)
-		context.doc.AddField(field)
+		flatVector, ok := processFlatVector(subVecI, dims)
+		if !ok {
+			return nil, false
+		}
+
+		copy(rv[i*dims:(i+1)*dims], flatVector)
```

Comment:
more of a question here, is there a difference (perf wise) between the one you've used and the following way of populating the float array:

`rv = append(rv, flatVector...)`

which is functionally same, since the copy statement is also technically appending to the rv slice.

i know that if you use the append way, it won't dynamically allocate the memory since you've already assigned it before the loop, but just curious as to whether one is better than the other. please let me know if this makes sense

Reply:
You are right that using copy() or append() can help us achieve the same functional goal.

With pre-allocating the space for resultant slice, we can ensure that append() doesn't perform any allocation while copying the elements.
But even then, it is slower than copy(). 
I am not sure of exact reason, but I suspect the book-keeping operations may be making append() slower than copy() in such case.

Here is my benchmarking test to compare these two:
https://gist.github.com/moshaad7/f459123d336a1a1375621fd5c19578a2